src/tcop/tcop.cpp

```diff
@@ -133,14 +126,255 @@ ResultType TrafficCop::ExecuteStatement(
+    // new txn, reset result status
+    curr_state.second = ResultType::SUCCESS;
+    txn = txn_manager.BeginTransaction(thread_id);
+    single_statement_txn_ = true;
+  } else {
+    // get ptr to current active txn
+    txn = curr_state.first;
+  }
+  // skip if already aborted
+  if (curr_state.second != ResultType::ABORTED) {
+    PL_ASSERT(txn);
+    p_status = executor::PlanExecutor::ExecutePlan(plan, txn, params, result,
+                                                   result_format);
+
+    if (p_status.m_result == ResultType::FAILURE) {
+      // only possible if init failed
+      init_failure = true;
+    }
+
+    auto txn_result = txn->GetResult();
+//    LOG_DEBUG("Txn Result Type %d", txn_result);
+
+    if (single_statement_txn_ == true || init_failure == true ||
+        txn_result == ResultType::FAILURE) {
+      LOG_TRACE(
+          "About to commit: single stmt: %d, init_failure: %d, txn_result: %s",
+          single_statement_txn, init_failure,
+          ResultTypeToString(txn_result).c_str());
+      switch (txn_result) {
+        case ResultType::SUCCESS:
+          // Commit single statement
+        LOG_TRACE("Commit Transaction");
+          p_status.m_result = CommitQueryHelper();
+          break;
+
+        case ResultType::FAILURE:
+        default:
+          // Abort
+        LOG_TRACE("Abort Transaction");
+          if (single_statement_txn_ == true) {
+            LOG_TRACE("Tcop_txn_state size: %lu", tcop_txn_state_.size());
+            p_status.m_result = AbortQueryHelper();
+          } else {
+            tcop_txn_state_.top().second = ResultType::ABORTED;
+            p_status.m_result = ResultType::ABORTED;
+          }
+      }
+    }
+  } else {
+    // otherwise, we have already aborted
+    p_status.m_result = ResultType::ABORTED;
+  }
+  LOG_TRACE("Check Tcop_txn_state Size After ExecuteStatementPlan %lu", tcop_txn_state_.size());
+  return p_status;
+}
+
+std::shared_ptr<Statement> TrafficCop::PrepareStatement(
+    const std::string &statement_name, const std::string &query_string,
+    UNUSED_ATTRIBUTE std::string &error_message,
+    const size_t thread_id UNUSED_ATTRIBUTE) {
+  LOG_TRACE("Prepare Statement name: %s", statement_name.c_str());
+  LOG_TRACE("Prepare Statement query: %s", query_string.c_str());
+
+  std::shared_ptr<Statement> statement(
+      new Statement(statement_name, query_string));
+  // transaction
+  // --multi-statements except BEGIN in a transaction
+  // ----not aborted
+  // ----COMMIT, ROLLBACK, and aborted
+  if (!tcop_txn_state_.empty()) {
+    single_statement_txn_ = false;
+    // multi-statment txn has been aborted, just block it
+    if (tcop_txn_state_.top().second == ResultType::ABORTED) {
+      return nullptr;
+    }
+  } else {
+    // --BEGIN
+    // --single-statement transaction
+    // ----prepareStatement abort
+    if (query_string == "BEGIN") {  // only begin a new transaction
```

Comment:
The packet manager is not supposed to parse the packet message content, and so it will not be capable to handle this.  That is my understanding of the Peloton to-be layered architecture, and let me know if I am wrong.   For this, I wonder whether you could do these transaction-related check-ups and settings after building the parse tree at line 327.  That will remove first this string comparison, and second let all the query content parsing done inside the parser.